Sources/GRPC/ConnectionManager.swift

```diff
@@ -413,77 +435,85 @@ internal final class ConnectionManager {
+      state.readyChannelMuxPromise.fail(GRPCStatus(code: .unavailable, message: nil))
+      state.candidateMuxPromise.fail(GRPCStatus(code: .unavailable, message: nil))
+      // In case we do successfully connect, close immediately.
+      state.candidate.whenSuccess {
+        $0.close(mode: .all, promise: nil)
+      }
+
+    // We have an active channel but the application doesn't know about it yet. We'll do the same
+    // as for `.connecting`.
+    case let .active(state):
+      shutdown = .shutdownByUser(closeFuture: self.eventLoop.makeSucceededFuture(()))
+      self.state = .shutdown(shutdown)
 
-        // Stop the creation of a new channel, if we can. If we can't then the task to
-        // `startConnecting()` will see our new `shutdown` state and ignore the request to connect.
-        state.scheduled.cancel()
+      // Fail the ready channel mux promise: we're shutting down so even if we manage to successfully
+      // connect the application shouldn't have access to the channel or multiplexer.
+      state.readyChannelMuxPromise.fail(GRPCStatus(code: .unavailable, message: nil))
+      // We have a channel, close it.
+      state.candidate.close(mode: .all, promise: nil)
 
-        // Fail the ready channel mux promise: we're shutting down so even if we manage to successfully
-        // connect the application shouldn't should have access to the channel.
-        state.readyChannelMuxPromise.fail(shutdown.reason)
+    // The channel is up and running: the application could be using it. We can close it and
+    // return the `closeFuture`.
+    case let .ready(state):
+      shutdown = .shutdownByUser(closeFuture: state.channel.closeFuture)
+      self.state = .shutdown(shutdown)
 
-      // We're already shutdown; nothing to do.
-      case let .shutdown(state):
-        shutdown = state
-      }
+      // We have a channel, close it.
+      state.channel.close(mode: .all, promise: nil)
 
-      return shutdown.closeFuture
+    // Like `.connecting` and `.active` the application does not have a `.ready` channel. We'll
+    // do the same but also cancel any scheduled connection attempts and deal with any fallout
+    // if we cancelled too late.
+    case let .transientFailure(state):
+      shutdown = .shutdownByUser(closeFuture: self.eventLoop.makeSucceededFuture(()))
+      self.state = .shutdown(shutdown)
+
+      // Stop the creation of a new channel, if we can. If we can't then the task to
+      // `startConnecting()` will see our new `shutdown` state and ignore the request to connect.
+      state.scheduled.cancel()
+
+      // Fail the ready channel mux promise: we're shutting down so even if we manage to successfully
+      // connect the application shouldn't should have access to the channel.
+      state.readyChannelMuxPromise.fail(shutdown.reason)
+
+    // We're already shutdown; nothing to do.
+    case let .shutdown(state):
+      shutdown = state
     }
+
+    return shutdown.closeFuture
   }
```

Comment:
This hasn't changed, it's just reindented thanks to the if-in-event-loop-else dance added just above.